### PR TITLE
Pare down TypeName by using specific types

### PR DESCRIPTION
**Pull request by theunrepentantgeek**
**What this PR does / why we need it**:

Continuing the refactoring and cleanup.

* Change return types of some methods on `InternalTypeName` to expose more functionality
* Change owner of `ResourceType` to an `InternalTypeName`
* Restrict `InternalTypeName` to using `InternalPackageReference`
* Remove unused methods from `TypeName` and `ExternalTypeName`
* Use `ExternalTypeName` for more external references
* Remove lots of casts that are now extraneous

**Prerequisites**

- [x] PR #3191 
- [x] PR #3192 

**How does this PR make you feel**:
gif

**If applicable**:
- [x] this PR contains tests

### CODE REVIEW

File: v2/tools/generator/internal/astmodel/type_visitor_builder.go [hunk 1]
```diff
@@ -426,5 +431,14 @@ func (b *TypeVisitorBuilder[C]) buildVisitInterfaceType() func(*TypeVisitor[C],
 		}
 	}
 
-	panic(fmt.Sprintf("unexpected InterfaceType func %#v", b.VisitInterfaceType))
+	panic(fmt.Sprintf("unexpected InterfaceType func %#v (%s)", b.VisitInterfaceType, b.diagnosticClue()))
+}
+
+func (b *TypeVisitorBuilder[C]) diagnosticClue() string {
```
Comment (matthchr) [inline v2/tools/generator/internal/astmodel/type_visitor_builder.go:437]: Its not immediately clear to me why `Caller(3)` in this function. Might help if it had a comment at the function level explaining what this function does/what its expected output is, and possibly a comment on the `Caller(3)` explaining whats special about 3.

File: v2/tools/generator/internal/codegen/pipeline/add_operator_spec.go [hunk 2]
```diff
@@ -389,7 +389,7 @@ func (b *operatorSpecBuilder) addSecretsToOperatorSpec(
 	secretsTypeName := resourceName.WithName(
 		b.idFactory.CreateIdentifier(
 			resourceName.Name()+"OperatorSecrets",
-			astmodel.Exported)).(astmodel.InternalTypeName)
+			astmodel.Exported))
```
Comment (matthchr) [inline v2/tools/generator/internal/codegen/pipeline/add_operator_spec.go:392]: yay!

File: v2/tools/generator/internal/codegen/pipeline/strip_unused_types_test.go [hunk 3]
```diff
@@ -6,6 +6,7 @@
 package pipeline
 
 import (
+	"github.com/Azure/azure-service-operator/v2/tools/generator/internal/test"
```
Comment (matthchr) [inline v2/tools/generator/internal/codegen/pipeline/strip_unused_types_test.go:9]: import ordering

File: v2/tools/generator/internal/codegen/storage/conversion_graph_builder.go [hunk 4]
```diff
@@ -43,7 +43,7 @@ func (b *ConversionGraphBuilder) Add(names ...astmodel.TypeName) {
 // AddAll includes the TypeNames in the supplied set in conversion graph
 func (b *ConversionGraphBuilder) AddAll(set astmodel.TypeNameSet) {
 	for name := range set {
-		b.Add(name)
+		b.Add(name.(astmodel.InternalTypeName))
```
Comment (matthchr) [inline v2/tools/generator/internal/codegen/storage/conversion_graph_builder.go:46]: awkward, does this go away?

**Review by matthchr — APPROVED** [review]

File: v2/tools/generator/internal/astmodel/external_type_name.go [hunk 5]
```diff
@@ -130,8 +77,8 @@ func (tn ExternalTypeName) RequiredPackageReferences() *PackageReferenceSet {
 }
 
 // Equals returns true if the passed type is the same TypeName, false otherwise
-func (tn ExternalTypeName) Equals(t Type, override EqualityOverrides) bool {
-	if tn == t && override.TypeName == nil {
+func (tn ExternalTypeName) Equals(t Type, _ EqualityOverrides) bool {
```
[the remark this answers is not in the document]
Comment (theunrepentantgeek) [reply v2/tools/generator/internal/astmodel/external_type_name.go:80]: Planning on changing the override in `EqualityOverrides` to be `InternalTypeName` specific; we can add an explicit `ExternalTypeName` override if needed, but we don't need it at the moment.

File: v2/tools/generator/internal/astmodel/type_visitor_builder.go (hunk 1 above)
Comment (theunrepentantgeek) [reply v2/tools/generator/internal/astmodel/type_visitor_builder.go:437]: I've added a detailed comment to explain.

File: v2/tools/generator/internal/codegen/pipeline/strip_unused_types_test.go (hunk 3 above)
Comment (theunrepentantgeek) [reply v2/tools/generator/internal/codegen/pipeline/strip_unused_types_test.go:9]: Fixed.

File: v2/tools/generator/internal/codegen/storage/conversion_graph_builder.go (hunk 4 above)
Comment (theunrepentantgeek) [reply v2/tools/generator/internal/codegen/storage/conversion_graph_builder.go:46]: Yes, in #3205.

**Pull request merged**